File: r_code/tests/testthat/test-node-alignment-and-seed.R
# Tests for node-universe alignment, and seed reproducibility
#
# Covers:
# 1. Unequal node-universe rejection across all fit methods
# 2. Seed reproducibility for community detection

# -----------------------------------------------------------------------
# 1. Unequal node-universe rejection
# -----------------------------------------------------------------------

test_that("different matrix sizes rejected by jaccard", {
  expect_error(
    fit_multilayer_jaccard(list(diag(4), diag(6)), algorithm = "louvain"),
    "same node"
  )
})

test_that("different matrix sizes rejected by overlap", {
  expect_error(
    fit_multilayer_overlap(list(diag(4), diag(6)), algorithm = "louvain"),
    "same node"
  )
})

test_that("different matrix sizes rejected by weighted_jaccard", {
  expect_error(
    fit_multilayer_weighted_jaccard(list(diag(4), diag(6)), algorithm = "louvain"),
    "same node"
  )
})

test_that("different matrix sizes rejected by weighted_overlap", {
  expect_error(
    fit_multilayer_weighted_overlap(list(diag(4), diag(6)), algorithm = "louvain"),
    "same node"
  )
})

test_that("different matrix sizes rejected by identity", {
  expect_error(
    fit_multilayer_identity_ties(list(diag(4), diag(6)), algorithm = "louvain"),
    "same node"
  )
})

test_that("three layers one different size rejected", {
  expect_error(
    fit_multilayer_jaccard(list(diag(5), diag(5), diag(4)), algorithm = "louvain"),
    "same node"
  )
})

test_that("same size matrices accepted", {
  fit <- fit_multilayer_jaccard(list(diag(5), diag(5)), algorithm = "louvain")
  expect_equal(length(fit$layer_communities), 2)
})


# -----------------------------------------------------------------------
# 2. Seed reproducibility
# -----------------------------------------------------------------------

make_test_layers <- function(seed = 42) {
  set.seed(seed)
  n <- 20
  memberships <- sample(1:3, n, replace = TRUE)
  layers <- lapply(1:3, function(l) {
    mat <- matrix(0, n, n)
    for (i in 1:(n - 1)) {
      for (j in (i + 1):n) {
        prob <- if (memberships[i] == memberships[j]) 0.4 else 0.05
        tie <- rbinom(1, 1, prob)
        mat[i, j] <- tie
        mat[j, i] <- tie
      }
    }
    mat
  })
  layers
}

test_that("seed produces identical results for jaccard", {
  layers <- make_test_layers()
  fit1 <- fit_multilayer_jaccard(layers, algorithm = "louvain", seed = 123)
  fit2 <- fit_multilayer_jaccard(layers, algorithm = "louvain", seed = 123)

  for (i in seq_along(fit1$layer_communities)) {
    expect_equal(
      as.integer(fit1$layer_communities[[i]]$membership),
      as.integer(fit2$layer_communities[[i]]$membership)
    )
  }
})

test_that("seed produces identical results for overlap", {
  layers <- make_test_layers()
  fit1 <- fit_multilayer_overlap(layers, algorithm = "louvain", seed = 456)
  fit2 <- fit_multilayer_overlap(layers, algorithm = "louvain", seed = 456)

  for (i in seq_along(fit1$layer_communities)) {
    expect_equal(
      as.integer(fit1$layer_communities[[i]]$membership),
      as.integer(fit2$layer_communities[[i]]$membership)
    )
  }
})

test_that("seed produces identical results for identity", {
  layers <- make_test_layers()
  fit1 <- fit_multilayer_identity_ties(layers, algorithm = "louvain", seed = 789)
  fit2 <- fit_multilayer_identity_ties(layers, algorithm = "louvain", seed = 789)

  for (i in seq_along(fit1$layer_communities)) {
    expect_equal(
      as.integer(fit1$layer_communities[[i]]$membership),
      as.integer(fit2$layer_communities[[i]]$membership)
    )
  }
})

test_that("no seed runs without error", {
  layers <- make_test_layers()
  fit <- fit_multilayer_jaccard(layers, algorithm = "louvain", seed = NULL)
  expect_equal(length(fit$layer_communities), 3)
})

test_that("simulate_and_fit_multilayer with seed is reproducible", {
  out1 <- simulate_and_fit_multilayer(n_nodes = 15, n_layers = 2, seed = 77)
  out2 <- simulate_and_fit_multilayer(n_nodes = 15, n_layers = 2, seed = 77)

  for (i in seq_along(out1$fit$layer_communities)) {
    expect_equal(
      as.integer(out1$fit$layer_communities[[i]]$membership),
      as.integer(out2$fit$layer_communities[[i]]$membership)
    )
  }
})

test_that("different named node sets rejected even at equal size", {
  m1 <- matrix(0, 3, 3, dimnames = list(c("a", "b", "c"), c("a", "b", "c")))
  m2 <- matrix(0, 3, 3, dimnames = list(c("a", "b", "d"), c("a", "b", "d")))
  m1["a", "b"] <- m1["b", "a"] <- 1
  m2["a", "b"] <- m2["b", "a"] <- 1
  expect_error(
    fit_multilayer_jaccard(list(m1, m2), algorithm = "louvain"),
    "same node"
  )
})

test_that("identity ties allow unequal node sets when opted in", {
  g1 <- igraph::graph_from_data_frame(
    data.frame(from = c("A", "A", "B"), to = c("B", "C", "C")),
    directed = FALSE
  )
  g2 <- igraph::graph_from_data_frame(
    data.frame(from = c("B", "B", "C"), to = c("C", "D", "D")),
    directed = FALSE
  )
  expect_error(
    fit_multilayer_identity_ties(list(g1, g2), algorithm = "louvain"),
    "same node"
  )
  fit <- fit_multilayer_identity_ties(
    list(g1, g2), algorithm = "louvain", allow_unequal_nodes = TRUE
  )
  ties <- fit$interlayer_ties
  expect_setequal(ties$node, c("B", "C"))
})

test_that("seed produces identical results for weighted methods", {
  layers <- make_test_layers()
  for (fn in list(fit_multilayer_weighted_jaccard, fit_multilayer_weighted_overlap)) {
    fit1 <- fn(layers, algorithm = "louvain", seed = 42)
    fit2 <- fn(layers, algorithm = "louvain", seed = 42)
    for (i in seq_along(fit1$layer_communities)) {
      expect_equal(
        as.integer(fit1$layer_communities[[i]]$membership),
        as.integer(fit2$layer_communities[[i]]$membership)
      )
    }
  }
})

test_that("seed produces identical results for hungarian", {
  layers <- make_test_layers()
  fit1 <- fit_multilayer_hungarian(layers, algorithm = "louvain", seed = 42)
  fit2 <- fit_multilayer_hungarian(layers, algorithm = "louvain", seed = 42)
  for (i in seq_along(fit1$meta_communities)) {
    expect_equal(
      as.integer(fit1$meta_communities[[i]]),
      as.integer(fit2$meta_communities[[i]])
    )
  }
})

test_that("seeded fit does not disturb the caller's RNG stream", {
  layers <- make_test_layers()
  set.seed(999)
  before <- .Random.seed
  invisible(fit_multilayer_jaccard(layers, algorithm = "louvain", seed = 42))
  expect_identical(.Random.seed, before)
  # and the caller's stream continues as if the fit never happened
  set.seed(999)
  draws_clean <- runif(5)
  set.seed(999)
  invisible(fit_multilayer_jaccard(layers, algorithm = "louvain", seed = 42))
  draws_after_fit <- runif(5)
  expect_identical(draws_clean, draws_after_fit)
})

File: python_code/tests/test_node_alignment_and_seed.py
"""
Tests for node-universe alignment, arbitrary node labels, and seed reproducibility.

Covers:
1. Unequal node-universe rejection across all fit methods
2. String and non-contiguous integer node labels via NetworkX graphs
3. Seed reproducibility for Louvain community detection
"""

import networkx as nx
import numpy as np
import pytest

from dynamic_multiplex import (
    fit_multilayer_identity_ties,
    fit_multilayer_jaccard,
    fit_multilayer_overlap,
    fit_multilayer_weighted_jaccard,
    fit_multilayer_weighted_overlap,
)

ALL_FIT_FNS = [
    fit_multilayer_jaccard,
    fit_multilayer_overlap,
    fit_multilayer_weighted_jaccard,
    fit_multilayer_weighted_overlap,
    fit_multilayer_identity_ties,
]


# -----------------------------------------------------------------------
# 1. Unequal node-universe rejection
# -----------------------------------------------------------------------


class TestUnequalNodeUniverseRejection:
    def test_different_matrix_sizes_rejected(self):
        """Layers with different matrix sizes should raise ValueError."""
        layer_small = np.eye(4)
        layer_large = np.eye(6)
        for fn in ALL_FIT_FNS:
            with pytest.raises(ValueError, match="same node"):
                fn([layer_small, layer_large], algorithm="louvain")

    def test_different_networkx_node_sets_rejected(self):
        """NetworkX graphs with different node sets should raise ValueError."""
        g1 = nx.Graph()
        g1.add_edges_from([(0, 1), (1, 2)])
        g2 = nx.Graph()
        g2.add_edges_from([(0, 1), (1, 3)])  # node 3 not in g1, node 2 not in g2

        for fn in ALL_FIT_FNS:
            with pytest.raises(ValueError, match="same node"):
                fn([g1, g2], algorithm="louvain")

    def test_same_size_matrices_accepted(self):
        """Layers with same-sized matrices should work fine."""
        layers = [np.eye(5), np.eye(5)]
        out = fit_multilayer_jaccard(layers, algorithm="louvain")
        assert len(out["layer_communities"]) == 2

    def test_three_layers_one_different_rejected(self):
        """Three layers where one differs should raise ValueError."""
        layers = [np.eye(5), np.eye(5), np.eye(4)]
        with pytest.raises(ValueError, match="same node"):
            fit_multilayer_jaccard(layers, algorithm="louvain")


# -----------------------------------------------------------------------
# 2. String and non-contiguous node labels
# -----------------------------------------------------------------------


class TestArbitraryNodeLabels:
    def _make_string_labeled_graph(self, labels, edges, weights=None):
        g = nx.Graph()
        g.add_nodes_from(labels)
        for i, (u, v) in enumerate(edges):
            w = weights[i] if weights else 1.0
            g.add_edge(u, v, weight=w)
        return g

    def test_string_node_labels_jaccard(self):
        """Graphs with string node labels should produce valid 1-indexed output."""
        labels = ["alice", "bob", "carol", "dave"]
        edges = [("alice", "bob"), ("bob", "carol"), ("carol", "dave"), ("alice", "carol")]
        g1 = self._make_string_labeled_graph(labels, edges)
        g2 = self._make_string_labeled_graph(labels, edges)

        out = fit_multilayer_jaccard([g1, g2], algorithm="louvain")
        for lc in out["layer_communities"]:
            assert len(lc.membership) == 4
            assert min(lc.membership.keys()) >= 1
            assert max(lc.membership.keys()) <= 4

    def test_string_labels_all_methods(self):
        """All fit methods should handle string-labeled graphs."""
        labels = ["x", "y", "z"]
        edges = [("x", "y"), ("y", "z")]
        g1 = self._make_string_labeled_graph(labels, edges)
        g2 = self._make_string_labeled_graph(labels, edges)

        for fn in ALL_FIT_FNS:
            out = fn([g1, g2], algorithm="louvain")
            assert len(out["layer_communities"]) == 2
            for lc in out["layer_communities"]:
                assert len(lc.membership) == 3

    def test_non_contiguous_integer_labels(self):
        """Graphs with non-contiguous integer labels (e.g., 10, 20, 30)."""
        g1 = nx.Graph()
        g1.add_nodes_from([10, 20, 30, 40])
        g1.add_edges_from([(10, 20, {"weight": 1.0}), (30, 40, {"weight": 1.0})])
        g2 = nx.Graph()
        g2.add_nodes_from([10, 20, 30, 40])
        g2.add_edges_from([(10, 20, {"weight": 1.0}), (30, 40, {"weight": 1.0})])

        out = fit_multilayer_jaccard([g1, g2], algorithm="louvain")
        for lc in out["layer_communities"]:
            assert len(lc.membership) == 4
            assert min(lc.membership.keys()) >= 1
            assert max(lc.membership.keys()) <= 4

    def test_identity_ties_with_string_labels(self):
        """Identity ties should produce correct node range with string labels."""
        labels = ["a", "b", "c", "d", "e"]
        edges = [("a", "b"), ("c", "d"), ("d", "e")]
        g1 = self._make_string_labeled_graph(labels, edges)
        g2 = self._make_string_labeled_graph(labels, edges)

        out = fit_multilayer_identity_ties([g1, g2], algorithm="louvain")
        ties = out["interlayer_ties"]
        # 5 nodes * 1 link = 5 rows
        assert ties.shape[0] == 5
        assert set(ties["node"]) == {1, 2, 3, 4, 5}

    def test_weighted_methods_with_string_labels(self):
        """Weighted methods should compute correct node strengths after remapping."""
        labels = ["p", "q", "r"]
        edges = [("p", "q"), ("q", "r")]
        weights = [3.0, 7.0]
        g1 = self._make_string_labeled_graph(labels, edges, weights)
        g2 = self._make_string_labeled_graph(labels, edges, weights)

        out = fit_multilayer_weighted_jaccard([g1, g2], algorithm="louvain", add_self_loops=False)
        assert "interlayer_ties" in out
        assert "weighting" in out
        assert out["weighting"] == "node_strength"


# -----------------------------------------------------------------------
# 3. Seed reproducibility
# -----------------------------------------------------------------------


class TestSeedReproducibility:
    def _make_layers(self, seed=42):
        rng = np.random.default_rng(seed)
        n = 30
        memberships = rng.integers(1, 4, size=n)
        layers = []
        for _ in range(3):
            mat = np.zeros((n, n))
            for i in range(n - 1):
                for j in range(i + 1, n):
                    prob = 0.4 if memberships[i] == memberships[j] else 0.05
                    tie = rng.binomial(1, prob)
                    mat[i, j] = tie
                    mat[j, i] = tie
            layers.append(mat)
        return layers

    def test_seed_produces_identical_results(self):
        """Same seed should produce identical community assignments."""
        layers = self._make_layers()

        out1 = fit_multilayer_jaccard(layers, algorithm="louvain", seed=123)
        out2 = fit_multilayer_jaccard(layers, algorithm="louvain", seed=123)

        for lc1, lc2 in zip(out1["layer_communities"], out2["layer_communities"]):
            assert lc1.membership == lc2.membership
            assert lc1.communities == lc2.communities

    def test_different_seeds_may_differ(self):
        """Different seeds should generally produce different results (not guaranteed
        but very likely for stochastic Louvain on non-trivial graphs)."""
        layers = self._make_layers()

        out1 = fit_multilayer_jaccard(layers, algorithm="louvain", seed=1)
        out2 = fit_multilayer_jaccard(layers, algorithm="louvain", seed=999)

        # Just check they run without error; exact equality is not guaranteed
        assert len(out1["layer_communities"]) == 3
        assert len(out2["layer_communities"]) == 3

    def test_seed_reproducibility_all_methods(self):
        """All fit methods should support seed parameter."""
        layers = self._make_layers()

        for fn in ALL_FIT_FNS:
            out1 = fn(layers, algorithm="louvain", seed=42)
            out2 = fn(layers, algorithm="louvain", seed=42)
            for lc1, lc2 in zip(out1["layer_communities"], out2["layer_communities"]):
                assert lc1.membership == lc2.membership

    def test_no_seed_runs_without_error(self):
        """Passing seed=None (default) should work fine."""
        layers = self._make_layers()
        out = fit_multilayer_jaccard(layers, algorithm="louvain", seed=None)
        assert len(out["layer_communities"]) == 3

    def test_seed_in_simulate_and_fit(self):
        """simulate_and_fit_multilayer with seed should be reproducible."""
        from dynamic_multiplex import simulate_and_fit_multilayer

        out1 = simulate_and_fit_multilayer(n_nodes=20, n_layers=2, seed=77)
        out2 = simulate_and_fit_multilayer(n_nodes=20, n_layers=2, seed=77)

        for lc1, lc2 in zip(
            out1["fit"]["layer_communities"], out2["fit"]["layer_communities"]
        ):
            assert lc1.membership == lc2.membership
